src/main/java/spoon/support/visitor/java/JavaReflectionVisitorImpl.java

```diff
@@ -325,12 +326,21 @@ public <T> void visitConstructor(Constructor<T> constructor) {
 		}
 	}
 
-	private int getNumberOfEnclosingClasses(Class<?> clazz) {
-		int depth = 0;
-		while (Modifier.isStatic(clazz.getModifiers()) == false && (clazz = clazz.getEnclosingClass()) != null) {
-			depth++;
-		}
-		return depth;
+	/**
+	 * Check whether the constructor parameter is implicit.
+	 * It is not enough to simply use {@link Parameter#isImplicit()} as the class file format before Java 8
+	 * has no way to embed the required information. As of Java 8, the information is stored in the
+	 * MethodParameters attribute, however javac does not emit the attribute by default in current Java versions.
+	 */
+	private boolean isImplicitParameter(Parameter parameter, Constructor<?> constructor, boolean isFirstParameter) {
+		if (parameter.isImplicit()) {
+			return true;
+		}
+		// best effort fallback
+		if (isFirstParameter && parameter.getType() == constructor.getDeclaringClass().getEnclosingClass()) {
+			return true;
+		}
+		return false;
```

Comment:
Optional comment: You could contract this into a single return statement.

```suggestion
		return isFirstParameter && parameter.getType() == constructor.getDeclaringClass().getEnclosingClass()
```

IMO it's better to just return like this, but I've met people who disagree so I'll leave it up to you if you want to change it or not.

Reply:
I applied the suggested change.